src/main/java/org/team1540/robot2018/motion/FollowProfile.java

```diff
@@ -0,0 +1,145 @@
+  private Trajectory right;
+  private boolean finished;
+
+  public FollowProfile(String profileName) {
+    DriveProfile profile = Robot.profiles.getProfile(profileName);
+    this.left = profile.getLeft();
+    this.right = profile.getRight();
+
+    requires(Robot.drivetrain);
+  }
+
+  public FollowProfile(Trajectory left, Trajectory right) {
+    this.left = left;
+    this.right = right;
+
+    requires(Robot.drivetrain);
+  }
+
+  private void run() {
+    //TODO: perform linear interpolation between trajectory points?
+    Segment leftSegment = getCurrentSegment(left, timer.get());
+    Segment rightSegment = getCurrentSegment(right, timer.get());
+
+    double robotHeading = 2 * PI - Math.toRadians(
+        Robot.navx.getYaw() < 0 ? 360 + Robot.navx.getYaw() : Robot.navx.getYaw());
+    double desiredHeading = leftSegment.heading;
+
+    double regularError = robotHeading - desiredHeading;
+    double differentError =
+        (2 * PI - max(robotHeading, desiredHeading)) - min(robotHeading, desiredHeading);
+
+    double headingError = abs(regularError) < abs(differentError) ? regularError : differentError;
+
+
+    // heading is negated for the left side only so that negative heading errors (i.e. too far right)
+    // result in the left side slowing but the right side speeding up
+    Robot.drivetrain.setLeft(ControlMode.Position, Tuning.drivetrainEncoderTPU
+        * leftSegment.position, getBump(leftSegment.acceleration, leftSegment.velocity, headingError));
+    Robot.drivetrain.setRight(ControlMode.Position, Tuning.drivetrainEncoderTPU
+        * rightSegment.position, getBump(rightSegment.acceleration, rightSegment.velocity, -headingError));
+  }
+
+  @Override
+  protected boolean isFinished() {
+    return finished;
+  }
+
+  // stolen from Jonathan's follower
+  private Segment getCurrentSegment(Trajectory trajectory, double currentTime) {
+    // Start from the current time and find the closest point.
+    int startIndex = Math.toIntExact(Math.round(currentTime / trajectory.segments[0].dt));
+
+    int length = trajectory.segments.length;
+    int index = startIndex;
+    if (startIndex >= length - 1) {
+      index = length - 1;
+      finished = true;
+    }
+    return trajectory.segments[index];
+  }
+
+  private Segment getInterpolatedSegment(Trajectory trajectory, double currentTime) {
+    // Start from the current time and find the closest point.
+    int startIndex = (int) Math.floor(currentTime / trajectory.segments[0].dt);
+    int endIndex = (int) Math.ceil(currentTime / trajectory.segments[0].dt);
+
+    int length = trajectory.segments.length;
+    if (endIndex >= length - 1) {
+      startIndex = length - 1;
+      endIndex = length - 1;
+      finished = true;
+    }
+
+    double timeSinceLower = currentTime % trajectory.segments[0].dt;
+    Segment lower = trajectory.segments[startIndex];
+    Segment upper = trajectory.segments[endIndex];
+
+    //noinspection SuspiciousNameCombination
```

Comment:
uh

Reply:
IntelliJ inspection, it was confused that I was passing the setpoint Y to a parameter called X

Reply:
Does it not use annotations?